include/SampleThumbnail.h

```diff
@@ -0,0 +1,146 @@
+   the visualization however (i.e., we are not reading from original sample data when drawing), this provides a
+   significant performance boost that wouldn't be possible otherwise.
+ */
+class LMMS_EXPORT SampleThumbnail
+{
+public:
+	struct VisualizeParameters
+	{
+		QRect sampleRect; //!< A rectangle that covers the entire range of samples.
+
+		QRect drawRect; //!< Specifies the location in `sampleRect` where the waveform will be drawn. Equals
+						//!< `sampleRect` when null.
+
+		QRect viewportRect; //!< Clips `drawRect`. Equals `drawRect` when null.
+
+		float amplification = 1.0f; //!< The amount of amplification to apply to the waveform.
+
+		float sampleStart = 0.0f; //!< Where the sample begins for drawing.
+
+		float sampleEnd = 1.0f; //!< Where the sample ends for drawing.
+
+		bool reversed = false; //!< Determines if the waveform is drawn in reverse or not.
+	};
+
+	SampleThumbnail() = default;
+	SampleThumbnail(const Sample& sample);
+	void visualize(VisualizeParameters parameters, QPainter& painter) const;
+
+private:
+	class Thumbnail
+	{
+	public:
+		struct Peak
+		{
+			Peak() = default;
+
+			Peak(float min, float max)
+				: min(min)
+				, max(max)
+			{
+			}
+
+			Peak(const SampleFrame& frame)
+				: min(std::min(frame.left(), frame.right()))
+				, max(std::max(frame.left(), frame.right()))
+			{
+			}
+
+			Peak operator+(const Peak& other) const { return Peak(std::min(min, other.min), std::max(max, other.max)); }
+			Peak operator+(const SampleFrame& frame) const { return *this + Peak{frame}; }
+
+			float min = std::numeric_limits<float>::max();
+			float max = std::numeric_limits<float>::min();
+		};
+
+		Thumbnail() = default;
+		Thumbnail(std::vector<Peak> peaks, double samplesPerPeak);
+		Thumbnail(const float* buffer, size_t size, size_t width);
+
+		Thumbnail zoomOut(float factor) const;
+
+		Peak& operator[](size_t index) { return m_peaks[index]; }
+		const Peak& operator[](size_t index) const { return m_peaks[index]; }
+
+		Peak* peaks() { return m_peaks.data(); }
+		const Peak* peaks() const { return m_peaks.data(); }
+
+		int width() const { return m_peaks.size(); }
+		double samplesPerPeak() const { return m_samplesPerPeak; }
+
+	private:
+		std::vector<Peak> m_peaks;
+		double m_samplesPerPeak = 0.0;
+	};
+
+	struct SampleThumbnailEntry
+	{
+		QString filePath;
+		QDateTime lastModified;
+
+		friend bool operator==(const SampleThumbnailEntry& first, const SampleThumbnailEntry& second)
+		{
+			return first.filePath == second.filePath && first.lastModified == second.lastModified;
+		}
+	};
+
+	struct Hash
+	{
+		std::size_t operator()(const SampleThumbnailEntry& entry) const noexcept { return qHash(entry.filePath); }
+	};
+
+	using ThumbnailCache = std::vector<Thumbnail>;
+	std::shared_ptr<ThumbnailCache> m_thumbnailCache = std::make_shared<ThumbnailCache>();
+
+	inline static std::unordered_map<SampleThumbnailEntry, std::shared_ptr<ThumbnailCache>, Hash> s_sampleThumbnailCacheMap;
```

Comment:
As Saker mentioned, this thumbnail class is a good candidate to become a resource associated with a UUID in the future if we add a resource management system that handles caching.